junkd/main.go

```diff
@@ -77,142 +86,205 @@ func main() {
+	}
+	dataShards, parityShards = uint8(ad), uint8(pd)
+	return nil
 }
 
-func formatBpsString(b int64, t time.Duration) string {
-	const units = "KMGTPE"
-	const factor = 1000
+func newLogger(level zap.AtomicLevel, path string) (*zap.Logger, error) {
+	encCfg := zap.NewProductionEncoderConfig()
+	encCfg.EncodeTime = zapcore.ISO8601TimeEncoder
+	encCfg.EncodeDuration = zapcore.MillisDurationEncoder
+	encCfg.EncodeLevel = zapcore.CapitalColorLevelEncoder // colorized console
 
-	time := t.Truncate(time.Second).Seconds()
-	if time <= 0 {
-		return "0.00 bps"
-	}
+	console := zapcore.NewCore(
+		zapcore.NewConsoleEncoder(encCfg),
+		zapcore.AddSync(zapcore.Lock(os.Stdout)),
+		level,
+	)
 
-	// calculate bps
-	speed := float64(b*8) / time
-
-	// short-circuit for < 1000 bits/s
-	if speed < factor {
-		return fmt.Sprintf("%.2f bps", speed)
+	if strings.TrimSpace(path) == "" {
+		return zap.New(console), nil
 	}
 
-	var i = -1
-	for ; speed >= factor; i++ {
-		speed /= factor
+	f, err := os.OpenFile(path, os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0o644)
+	if err != nil {
+		return nil, err
 	}
-	return fmt.Sprintf("%.2f %cbps", speed, units[i])
-}
+	fileCore := zapcore.NewCore(
+		zapcore.NewJSONEncoder(encCfg),
+		zapcore.AddSync(f),
+		level,
+	)
 
-func printUploadSpeeds(ctx context.Context, log *zap.Logger) {
-	t := time.NewTicker(2 * time.Minute)
-	defer t.Stop()
-
-	for {
-		select {
-		case <-ctx.Done():
-			return
-		case <-t.C:
-			elapsedMu.Lock()
-			if len(elapsed) > 1000 {
-				elapsed = elapsed[len(elapsed)-1000:]
-			}
-			times := elapsed
-			elapsedMu.Unlock()
+	return zap.New(zapcore.NewTee(console, fileCore)), nil
+}
 
-			var avg time.Duration
-			if len(times) == 0 {
-				avg = time.Second
-			} else {
-				for _, t := range times {
-					avg += t
-				}
-				avg /= time.Duration(len(times))
-			}
-			log.Info("average upload time", zap.String("averageSpeed", formatBpsString(int64(redundantSlabSize), avg)))
-		}
+func humanReadableSize(bytes int64) string {
+	const unit = 1024
+	if bytes < unit {
+		return fmt.Sprintf("%d B", bytes)
 	}
+	div, exp := int64(unit), 0
+	for n := bytes / unit; n >= unit; n /= unit {
+		div *= unit
+		exp++
+	}
+	return fmt.Sprintf("%.1f %ciB", float64(bytes)/float64(div), "KMGTPE"[exp])
+}
+
+func parseFlags() config {
+	var cfg config
+	flag.StringVar(&cfg.indexerURL, "indexer.url", "http://localhost:9982", "the URL of the indexer API")
+	flag.StringVar(&cfg.appSecret, "app.secret", "", "a secret used to derive the application key")
+	flag.TextVar(&cfg.logLevel, "log.level", zap.NewAtomicLevelAt(zap.InfoLevel), "the log level to use")
+	flag.StringVar(&cfg.logPath, "log.path", "", "the path to write the log to")
+	flag.IntVar(&cfg.threads, "threads", 1, "the number of upload threads")
+	flag.IntVar(&cfg.slabs, "slabs", 1, "the number of slabs to upload")
+	flag.StringVar(&cfg.redundancyStr, "redundancy", "2-4", "the redundancy configuration")
```

Comment:
It's not obvious from the docstring whether this is 2+4 or 2 out of 4. Since you are already updating this might as well clarify.

Also I'd change it to our 10 out of 30 default redundancy to more accurately reflect the config most people use.